// dve/display/title.js
// Generated by CoffeeScript 1.9.3

/*

Add a title to your visualisations.

- type: title
  text: Wind Speed (kts)
 */
var calculate_layout, d3;

d3 = require('d3');

calculate_layout = (function(_this) {
  return function(dimensions) {
    var canvas, margin;
    margin = {
      top: 0,
      right: 0,
      bottom: 0,
      left: 0
    };
    dimensions = {
      width: dimensions[0],
      height: 25
    };
    canvas = {
      top: margin.top,
      right: margin.right,
      bottom: margin.bottom,
      left: margin.left,
      width: dimensions.width - margin.left - margin.right,
      height: dimensions.height - margin.top - margin.bottom
    };
    return {
      margin: margin,
      dimensions: dimensions,
      canvas: canvas
    };
  };
})(this);

module.exports = function(spec, components) {
  var svg, title;
  svg = null;
  return title = {
    render: function(dom, state, params) {
      var layout;
      layout = calculate_layout(params.dimensions);
      svg = d3.select(dom).append('svg').attr('class', 'item title');
      svg.append('g').attr('class', 'title').attr('transform', "translate(" + layout.canvas.left + "," + layout.canvas.top + ")").append('text').attr('class', 'infotext').attr('dy', 20).text(spec.text);
      return title.resize(params.dimensions);
    },
    resize: function(dimensions) {
      var layout;
      layout = calculate_layout(dimensions);
      return svg.attr('width', layout.dimensions.width).attr('height', layout.dimensions.height);
    }
  };
};
